usr/src/uts/common/fs/zfs/zil.c

```diff
@@ -908,9 +918,13 @@ zil_lwb_write_init(zilog_t *zilog, lwb_t *lwb)
 		    ZIO_FLAG_CANFAIL);
 	}
 	if (lwb->lwb_zio == NULL) {
+		if (zilog->zl_cur_used <= zil_slog_limit || !lwb->lwb_slog)
+			prio = ZIO_PRIORITY_SYNC_WRITE;
+		else
+			prio = ZIO_PRIORITY_ASYNC_WRITE;
```

Comment:
In general, slog writes tend to be ordered with few long seeks, though there can be many short retrogrades.  In any case, today there is no reason to try and optimize slog for HDDs. 

However, I fail to see the advantage here. A slog is its own vdev so it sees no scheduler contention with other I/O types, especially reads. For now, it is not uncommon to need much higher concurrency for SSDs than the default 10, and so it gets tuned [*]. Perhaps a more general solution is to add a separate slog-based priority? In other words, this seems hackish and counter-intuitive to someone trying to tune their system for high throughput with slogs -- a large portion of the installed base including anyone who runs in HA cluster mode.

Similarly, sorting offers no benefit and simply consumes CPU cycles, so taking the async priority to get past the max limit of sync priority seems marginal at best.

Finally, it is not uncommon to see workloads where slogs must be striped to get more throughput and logbias=throughput is not a viable option. We expect this to be fixed in hardware as NVMe/NVMf become more common, but for SAS/SATA/FC we're stuck in a world where the networks are faster than the disk interconnect and disks.

Also, can you explain how this is expected to work with large recordsize?

[*] many flash SSDs cannot reach full write throughput with only 10 concurrent I/Os.